Research/ResearchUI/iOS/Views/RSDVerticalGridCollectionView.swift

```diff
@@ -0,0 +1,191 @@
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+import Foundation
+
+/// The 'RSDVerticalGridCollectionView' is a normal collection view that forces a
+/// 'RSDVerticalGridCollectionViewFlowLayout' as the colelction view layout
+open class RSDVerticalGridCollectionView: UICollectionView {
+    
+    /// Force cast of collectin view layout as a RSDVerticalGridCollectionViewFlowLayout
+    open var gridLayout: RSDVerticalGridCollectionViewFlowLayout {
+        return self.collectionViewLayout as! RSDVerticalGridCollectionViewFlowLayout
+    }
+    
+    public override init(frame: CGRect, collectionViewLayout layout: UICollectionViewLayout) {
+        
+        super.init(frame: frame, collectionViewLayout: layout)
+                        
+        if !layout.isKind(of: RSDVerticalGridCollectionViewFlowLayout.self) {
+            debugPrint("WARNING: UICollectionViewLayout must be RSDVerticalGridCollectionViewFlowLayout, provided one will be overwritten")
+            self.initDefaultFlowLayout()
+        }
+    }
+
+    public required init?(coder: NSCoder) {
+        super.init(coder: coder)
+        self.initDefaultFlowLayout()
+    }
+    
+    fileprivate func initDefaultFlowLayout() {
+        let layout = RSDVerticalGridCollectionViewFlowLayout()
+        self.collectionViewLayout = layout
+    }
+}
+
+/// The 'RSDVerticalGridCollectionViewFlowLayout' creates a centered grid layout in conjunction with
+/// the 'RSDVerticalGridCollectionView'.  As the class name implies, this class only works currently in a vertical orientation.
+open class RSDVerticalGridCollectionViewFlowLayout: UICollectionViewFlowLayout {
+            
+    /// The number of columns in each row of the grid
+    open var columnCount: Int = 3
+    
+    /// The width of a cell in the row when it is full at the column count
+    fileprivate var cellWidth: CGFloat {
+        let collectionViewWidth = (collectionView?.bounds.width ?? 0)
+        if collectionViewWidth == 0 {
+            return 0
+        }
+        let width = (collectionViewWidth - (CGFloat(columnCount) * horizontalCellSpacing)) / CGFloat(columnCount)
+        // Round down the value so that we make sure the cells fit in the full width
+        return CGFloat(Int(width))
+    }
+        
+    /// The height of each individual cell
+    open var cellHeightAbsolute: CGFloat = 120
+    /// If non-negative, this will be used  to determine height instead of the absolute height
+    /// Ratio is width * ratio = height
+    open var cellHeightRatio: CGFloat = -1
```

Comment:
Why is this obfuscating the code?  It's a feature of the layout that is supported and tested.  It will help in several scenarios and will scale better on large layouts than the absolute height.

Reply:
I didn't see anywhere that you are testing this? If so, cool. Since I didn't see this referenced anywhere, didn't see a point in including it. Sorry.